In [1]:
# This code cell starts the necessary setup for Hour of CI lesson notebooks.
# First, it enables users to hide and unhide code by producing a 'Toggle raw code' button below.
# Second, it imports the hourofci package, which is necessary for lessons and interactive Jupyter Widgets.
# Third, it helps hide/control other aspects of Jupyter Notebooks to improve the user experience
# This is an initialization cell
# It is not displayed because the Slide Type is 'Skip'

from IPython.display import HTML, IFrame, Javascript, display
from ipywidgets import interactive
import ipywidgets as widgets
from ipywidgets import Layout

import getpass # This library allows us to get the username (User agent string)

# import package for hourofci project
import sys
sys.path.append('../../supplementary') # relative path (may change depending on the location of the lesson notebook)
import hourofci

# load javascript to initialize/hide cells, get user agent string, and hide output indicator
# hide code by introducing a toggle button "Toggle raw code"
HTML(''' 
    <script type="text/javascript" src=\"../../supplementary/js/custom.js\"></script>
    
    <style>
        .output_prompt{opacity:0;}
    </style>
    
    <input id="toggle_code" type="button" value="Toggle raw code">
''')

## Introduction to Geospatial Data
### Part 3 of 5
# Exploring Direct Georeferencing


## Reminder
<a href="#/slide-2-0" class="navigate-right" style="background-color:blue;color:white;padding:8px;margin:2px;font-weight:bold;">Continue with the lesson</a>

<font size="+1">

By continuing with this lesson you are granting your permission to take part in this research study for the Hour of Cyberinfrastructure: Developing Cyber Literacy for GIScience project. In this study, you will be learning about cyberinfrastructure and related concepts using a web-based platform that will take approximately one hour per lesson. Participation in this study is voluntary.

Participants in this research must be 18 years or older. If you are under the age of 18 then please exit this webpage or navigate to another website such as the Hour of Code at https://hourofcode.com, which is designed for K-12 students.

If you are not interested in participating please exit the browser or navigate to this website: http://www.umn.edu. Your participation is voluntary and you are free to stop the lesson at any time.

For the full description please navigate to this website: <a href="gateway-1.ipynb">Gateway Lesson Research Study Permission</a>.

</font>

## Let’s try using descriptions to identify location.
Drag each symbol to its correct location as shown in the table.

{{IFrame("supplementary/grid-1.html", width=970, height=530)}}

### Let's use numeric coordinates this time.
Drag each symbol to its correct location as shown in the table. This may be a little less easy, but use the same thinking as before!

{{IFrame("supplementary/grid-2.html", width=970, height=530)}}

### Coordinate systems are everywhere
They are used to label the location of things, so that we can find them again.

This dresser, for example, has a coordinate system!
- Your stuffed hippo is in the lower right drawer, go get it!
- Stuffed hippo **coordinates**: (lower, right drawer)

{{IFrame("supplementary/dresser.html", width=620, height=630)}}

This library too, has a coordinate system!
- The book about hippos you want to read is in aisle 1A on the upper shelf. Go get it!
- Hippo book **coordinates**: (1A aisle, upper shelf)

{{IFrame("supplementary/bookshelf.html", width=620, height=430)}}

Some **coordinate systems** are used so frequently that they are given dedicated names.

The **“Cartesian” coordinate system**, for example, was thought of by René Descartes. You learned about this in your math classes in grade school.

### Let’s try using the Cartesian coordinate system to draw a trajectory (path).

### Someone recorded the path of the hippo across this map.
Use the Cartesian coordinate system to draw the trajectory (path) of the hippo.
- The hippo's trajectory coordinates are: `[(-4,6), (0,5), (3,-2)]`

{{IFrame("supplementary/cartesian.html", width=620, height=540)}}

## <center> But the earth is a sphere!
<center><img src='supplementary/globe.jpg' alt='globe' style="width:300px;"> 
<small>Photo by <a href="https://unsplash.com/@bill_oxford">Bill Oxford</a> on <a href="https://unsplash.com/s/photos/globe">Unsplash</a></small>

In order to find coordinates on the spherical earth we use an <i>angular</i> measurement system that produces the coordinates *latitude and longitude*. 
<table>
    <tr style="background: #fff">
        <td><img src='supplementary/globelatlong.jfif' alt='Latitude and Longitude'></td> 
        <td><large>Latitude is the angle north or south of the plane of the equator.<br><br>Longitude is the angle east and west of a standard meridian (a line running pole to pole).</td>
    </tr>
</table>

<small>Image source <a href="https://openstax.org/details/books/calculus-volume-3">openstax</a></small>

This map shows the World Geodetic System (WGS for short). It is the coordinate system used to locate things on the Earth. It is what you get when you ask for coordinates in Google Maps and often what you get from a GPS.
<table>
    <tr style="background: #fff">
        <td width=60%> {{IFrame("supplementary/geodetic.html", width=650, height=530)}}</td>
        <td><large>The first number in WGS coordinates is the <i>latitude</i> - measured north and south from the equator.<br><br>The second number in WGS coordinates is the <i>longitude</i> - measured east and west from the prime meridian (runs through Greenwich England).</td>
    </tr>
</table>

### Find the location of hippos on this world map!

- There is a hippo family at coordinates (30°E, 15°N). Click on that location.<p> 

{{IFrame("supplementary/geodetic.html", width=980, height=530)}}

Now, let's look at the geographic references in that cities data again.<p>
On the next slide, remember to click the arrow left of the code chunk to run it. 

In [8]:
import geopandas
cities = geopandas.read_file("https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/110m/cultural/ne_110m_populated_places_simple.zip")
cities.head()
#this will take a while so be patient. Wait for the asterisk to change to a number.

scalerank  natscale  labelrank           featurecla          name namepar  \
0          8        10          3      Admin-0 capital  Vatican City    None   
1          7        20          0      Admin-0 capital    San Marino    None   
2          7        20          0      Admin-0 capital         Vaduz    None   
3          6        30          8  Admin-0 capital alt       Lobamba    None   
4          6        30          8      Admin-0 capital    Luxembourg    None   

  namealt  diffascii     nameascii  adm0cap  ...  rank_max rank_min  \
0    None          0  Vatican City      1.0  ...         2        2   
1    None          0    San Marino      1.0  ...         7        7   
2    None          0         Vaduz      1.0  ...         7        5   
3    None          0       Lobamba      0.0  ...         5        4   
4    None          0    Luxembourg      1.0  ...         9        8   

   geonameid  meganame       ls_name ls_match checkme min_zoom       ne_id  \
0  6691831.0      None  Vatican City        1       0      7.0  1159127243   
1  3168070.0      None    San Marino        1       5      6.1  1159146051   
2  3042030.0      None         Vaduz        1       0      6.7  1159146061   
3   935048.0      None       Lobamba        1       5      6.0  1159146343   
4  2960316.0      None    Luxembourg        1       0      6.0  1159146437   

                     geometry  
0   POINT (12.45339 41.90328)  
1   POINT (12.44177 43.93610)  
2    POINT (9.51667 47.13372)  
3  POINT (31.20000 -26.46667)  
4    POINT (6.13000 49.61166)  

[5 rows x 39 columns]

Back in that table, in the first row, rightmost column, you'll see that Vatican City is listed with the point geometry (12.45339 41.90328).<p>
<small><i>Hint: To go back one slide, click the '<' symbol in the bottom right.</i></small><p>
This is 12.45 degrees east of the prime meridian (E longitude) and 41.90 degrees north of the equator (N latitude). Thus the Vatican is a city in the middle of Europe.<p>
<font color = red>HOWEVER, and this is IMPORTANT!</font> We usually would state its WGS coordinate as<br>(41.9 N, 12.45 E), stating the latitude value first.<br><p><p>
While we commonly say WGS coodinates as the pair (lat long), when we record a geometric coordinate, we always state the horizontal distance before the vertical distance. That's why the table shows the location as (12.45339 41.90328).

## Assessment

Now, look at the geometry for Lobamba (4th row in the table two slides back). Answer the following questions.

What is the longitude? (first number, report only the full number before decimal)

In [5]:
text1 = widgets.Text(placeholder='Type your answer here...')
# Display widget
display(text1)

# Output function
def out1():
    print("The correct answer is 31.")
    
# Submit button
hourofci.SubmitBtn(user_agent, lesson, lesson_level, "4H", text1, out1)

Text(value='', placeholder='Type your answer here...')

Button(description='Submit', icon='check', layout=Layout(height='auto', width='auto'), style=ButtonStyle())

Output()

What is the latitude? (second number, report only the full number)

In [6]:
text2 = widgets.Text(placeholder='Type your answer here...')
# Display widget
display(text2)

# Output function
def out2():
    print("The correct answer is -26.\n")
    print("The negative latitude value indicates a location south of the equator.")
    
# Submit button
hourofci.SubmitBtn(user_agent, lesson, lesson_level, "4I", text2, out2)

Text(value='', placeholder='Type your answer here...')

Button(description='Submit', icon='check', layout=Layout(height='auto', width='auto'), style=ButtonStyle())

Output()

What continent is this city on?

In [7]:
text3 = widgets.Text(placeholder='Type your answer here...')
# Display widget
display(text3)

# Output function
def out3():
    print("The correct answer is Africa.")
    
# Submit button
hourofci.SubmitBtn(user_agent, lesson, lesson_level, "4J", text3, out3)

Text(value='', placeholder='Type your answer here...')

Button(description='Submit', icon='check', layout=Layout(height='auto', width='auto'), style=ButtonStyle())

Output()

Ok, you've mastered direct georeferencing - identifing the location of anything using a coordinate system. 

Let's take a look at indirect georeferencing next.

<a href="gd-5.ipynb">Click here to move on.</a>